### PR TITLE
sys/net/nanocoap: fix buffer overflow in separate response handling 

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

When RFC 8974 support (module `nanocoap_token_ext`) is in use, the request token may be longer than the buffer in the separate response context is large. This adds a check to not overflow the buffer.

Sadly, this is an API change: Preparing the separate response context can actually fail, so we need to report this with a return value.

The example application has been adapted to only proceed if the separate reply context could have been prepared, and rather directly emit a reset message if the token exceeds the static buffer.

### Testing procedure

A second commit has been sneaked in to add the `client_token` shell command to `tests/net/nanocoap_cli` that allow specifying the Token as hex. The most recently set token is used in `client`.

#### Preparation

```
$  sudo ./dist/tools/tapsetup/tapsetup 
```

#### Running the Server

This should be done with `master` and this PR

```
$ USEMODULE=nanocoap_token_ext make BOARD=native64 -C examples/nanocoap_server && examples/nanocoap_server/bin/native64/nanocoap_server.elf -w tap1
[...]
main(): This is RIOT! (Version: 2025.01-devel-261-g75828)
RIOT nanocoap example application
Waiting for address autoconfiguration...
{"IPv6 addresses": ["<IPV6_ADDR>"]}
```

copy the `<IPV6_ADDR>` part of the output

#### Running the Client

This should be done with this PR only (or cherry-picking the second commit onto `master`).

```
$ USEMODULE=nanocoap_token_ext make BOARD=native64 -C tests/net/nanocoap_cli flash term
[...]
2024-12-11 18:10:58,750 # RIOT native interrupts/signals initialized.
2024-12-11 18:10:58,751 # RIOT native board initialized.
2024-12-11 18:10:58,751 # RIOT native hardware initialization complete.
2024-12-11 18:10:58,751 # 
2024-12-11 18:10:58,752 # main(): This is RIOT! (Version: 2025.01-devel-262-g562036-sys/net/nanocoap/buffer-overflow-separate-response)
2024-12-11 18:10:58,752 # nanocoap test app
2024-12-11 18:10:58,752 # All up, running the shell now
> client_token deadbeefdeadbeefdeadbeef
2024-12-11 18:12:11,060 # client_token deadbeefdeadbeefdeadbeef
> client get fe80::24de:d0ff:fe85:bdb2 5683 /separate
2024-12-11 18:12:27,736 # client get fe80::24de:d0ff:fe85:bdb2 5683 /separate
2024-12-11 18:12:27,736 # nanocli: sending msg ID 1, 25 bytes
2024-12-11 18:13:45,417 # nanocli: msg send failed: -110
```

#### Output

On `master`, this will result in

```
RIOT native interrupts/signals initialized.
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2025.01-devel-261-g75828)
RIOT nanocoap example application
Waiting for address autoconfiguration...
{"IPv6 addresses": ["fe80::24de:d0ff:fe85:bdb2"]}
_separate_handler(): send ACK, schedule response
_separate_handler(): send delayed response
*** stack smashing detected ***: terminated
[1]    367561 IOT instruction (core dumped)  examples/nanocoap_server/bin/native64/nanocoap_server.elf -w tap1
```

With this PR, a RST message is replied and no buffer overflow / crash should appear.

### Issues/PRs references

None

### CODE REVIEW

File: tests/net/nanocoap_cli/nanocli_client.c [hunk 1]
```diff
@@ -164,9 +172,69 @@ static int _cmd_client(int argc, char **argv)
...
+            puts("Token invalid: Must be given in hex");
+            return 1;
+        }
+
+        if (!_get_nibble(&low, argv[1][(pos << 1) + 1])) {
+            puts("Token invalid: Must be given in hex");
+            return 1;
+        }
+
+        _client_token[pos] = (high << 4) | low;
```
Comment (benpicco) [inline tests/net/nanocoap_cli/nanocli_client.c:230]: Maybe move that to `scn_buf_hex(const char *str, size_t str_len, uint8_t *out, size_t out_len)`

File: tests/net/nanocoap_cli/nanocli_client.c [hunk 2]
```diff
@@ -164,9 +172,69 @@ static int _cmd_client(int argc, char **argv)
...
+    }
+
+    return 0;
+}
+SHELL_COMMAND(client_token, "Set Token for CoAP client", _cmd_client_token);
+
+
```
Comment (benpicco) [inline tests/net/nanocoap_cli/nanocli_client.c:237]:
```suggestion
```

File: sys/picolibc_syscalls_default/syscalls.c [hunk 3]
```diff
@@ -249,8 +249,8 @@ FILE picolibc_stdio =
 #endif
 
 FILE *const stdin = &picolibc_stdio;
-STDIO_ALIAS(stdout);
-STDIO_ALIAS(stderr);
+__attribute__((used)) STDIO_ALIAS(stdout);
+__attribute__((used)) STDIO_ALIAS(stderr);
```
[the remark this answers is not in the document]
Comment (maribu) [reply sys/picolibc_syscalls_default/syscalls.c:253]: Oh, that is unrelated. It fixes LTO with picolibc

File: tests/net/nanocoap_cli/nanocli_client.c (hunk 1 above)
Comment (maribu) [reply tests/net/nanocoap_cli/nanocli_client.c:231]: Done

File: tests/net/nanocoap_cli/nanocli_client.c [hunk 4]
```diff
@@ -164,9 +173,68 @@ static int _cmd_client(int argc, char **argv)
...
+        return true;
+    }
+
+    if (((uint8_t)'A' <= c) && (c <= (uint8_t)'F')) {
+        *dest = c - (uint8_t)'A' + 10;
+        return true;
+    }
+
+    return false;
+}
```
Comment (maribu) [inline tests/net/nanocoap_cli/nanocli_client.c:197]: That rebase worked super well :cry:

**Review by nandojve — APPROVED** [review]

**Comment by maribu** [comment]
Thx a lot!

**Pull request merged**